base/bucket.go

```diff
@@ -128,6 +179,98 @@ func (b BucketSpec) GetViewQueryTimeout() time.Duration {
 
 }
 
+// TLSConnect method is passed to cbdatasource, to be used when creating a TLS-enabled memcached connection.
+// Establishes a connection, then wraps w/ gomemcached Client for use by cbdatasource.
+// Will include client cert (x.509) authentication when specified in the BucketSpec.
+func (b BucketSpec) TLSConnect(prot, dest string) (rv *memcached.Client, err error) {
+
+	d := net.Dialer{
+		Deadline: time.Now().Add(30 * time.Second),
+	}
+
+	host, port, err := SplitHostPort(dest)
+	if err != nil {
+		return nil, err
+	}
+
+	port = fmt.Sprintf("%d", b.KvTLSPort)
+	dest = host + ":" + port
+
+	Infof(KeyAll, "Establishing TLS connection for DCP to destination %s", dest)
+
+	conn, err := d.Dial("tcp", dest)
+	if err != nil {
+		return nil, err
+	}
+
+	tcpConn, isTcpConn := conn.(*net.TCPConn)
+	if !isTcpConn {
+		return nil, fmt.Errorf("Unable to convert connection to TCPConn during DCP TLS connection (Connection type:%T)", conn)
+	} else {
+		err = tcpConn.SetNoDelay(false)
+		if err != nil {
+			return nil, pkgerrors.Wrapf(err, "Error setting NoDelay on tcpConn during TLS Connect")
+		}
+
+		tlsConfig := &tls.Config{}
+		if b.Certpath != "" && b.Keypath != "" {
+			var configErr error
+			tlsConfig, configErr = TLSConfigForX509(b.Certpath, b.Keypath, b.CACertPath)
+			if configErr != nil {
+				return nil, pkgerrors.Wrapf(configErr, "Error adding x509 to TLSConfig for DCP TLS connection")
+			}
+		}
+		tlsConfig.ServerName = host
+
+		tlsConn := tls.Client(tcpConn, tlsConfig)
```

Comment:
if `b.Certpath == '' && b.Keypath == ""` the code will make it here, but does it have any chance of succeeding?  Just wondering if it should even try to create a `tls.Client()` in that case.

Reply:
This code adds general TLS support for DCP/cbdatasource, not just the x.509 auth case.

Reply:
Ok now it makes more sense.